File: backend/routes/member.js
import express from "express";
const router = express.Router({ mergeParams: true });

import * as auth from "../services/auth.js";
import * as member from "../services/member.js";

import { api } from "../utils/controller-api.js";

// get all members
router.get("/", auth.isAuthorized, async function (req, res) {
  // @ts-ignore
  const nwid = req.params.nwid;
  api
    .get("controller/network/" + nwid + "/member")
    .then(async function (controllerRes) {
      const mids = Object.keys(controllerRes.data);
      const data = await member.getMembersData(nwid, mids);
      res.send(data);
    })
    .catch(function (err) {
      res.status(404).send({ error: `Network not found ${err}` });
    });
});

// get member
router.get("/:mid", auth.isAuthorized, async function (req, res) {
  // @ts-ignore
  const nwid = req.params.nwid;
  const mid = req.params.mid;
  const data = await member.getMembersData(nwid, [mid]);
  if (data[0]) {
    res.send(data[0]);
  } else {
    res.status(404).send({ error: "Member not found" });
  }
});

// update member
router.post("/:mid", auth.isAuthorized, async function (req, res) {
  // @ts-ignore
  const nwid = req.params.nwid;
  const mid = req.params.mid;
  member.updateMemberAdditionalData(nwid, mid, req.body);
  if (req.body.config) {
    api
      .post("controller/network/" + nwid + "/member/" + mid, req.body.config)
      .then(async function () {
        const data = await member.getMembersData(nwid, [mid]);
        res.send(data[0]);
      })
      .catch(function (err) {
        res.status(500).send({ error: err.message });
      });
  } else {
    const data = await member.getMembersData(nwid, [mid]);
    res.send(data[0]);
  }
});

// delete member
router.delete("/:mid", auth.isAuthorized, async function (req, res) {
  // @ts-ignore
  const nwid = req.params.nwid;
  const mid = req.params.mid;
  member.deleteMemberAdditionalData(nwid, mid);
  api
    .delete("controller/network/" + nwid + "/member/" + mid)
    .then(function () {})
    .catch(function (err) {
      res.status(500).send({ error: err.message });
    });
  // Need this to fix ZT controller bug https://github.com/zerotier/ZeroTierOne/issues/859
  const defaultConfig = {
    authorized: false,
    ipAssignments: [],
    capabilities: [],
    tags: [],
  };
  api
    .post("controller/network/" + nwid + "/member/" + mid, defaultConfig)
    .then(function (controllerRes) {
      res.status(controllerRes.status).send("");
    })
    .catch(function (err) {
      res.status(500).send({ error: err.message });
    });
});

export default router;

File: backend-py/routes/member.py
from flask import jsonify, request, Blueprint
from untils.response import Response
import sys, os
sys.path.append(os.getcwd())
from constants.constants import *

from services.member import MEMBER
services_member = MEMBER()

from vars import DARWIN, LINUX
sys_vars = DARWIN() if os.uname().sysname == "Darwin" else LINUX()

route_member = Blueprint('member', __name__)


@route_member.route('/api/network/<string:nwid>/member', methods=['GET', 'DELETE'])
def api_network_nwid_member(nwid : str):
    if request.method == "GET":
        
        mids = zerotier_api.make_request(request.method, f"/controller/network/{nwid}/member")
        
        data = services_member.getMembersData(nwid, mids.json().keys())
        
        response_object = Response(200, "ok", data)
        
    return jsonify(response_object)


@route_member.route('/api/network/<string:nwid>/member/<string:mid>', methods=['DELETE', 'POST'])
def api_network_nwid_member_member(nwid : str, mid : str):
    if request.method == "POST":

        services_member.updateMemberAdditionalData(nwid, mid, request.json)

        if type(request.json) == list:
            zerotier_api.make_request(request.method, f"/controller/network/{nwid}/member/{mid}", json = {"authorized" : request.json[0]["authorized"]})
            data = services_member.getMembersData(nwid, [mid])

        elif (request.json.get("config")):

            zerotier_api.make_request(request.method, f"/controller/network/{nwid}/member/{mid}", json = request.json["config"])
            data = services_member.getMembersData(nwid, [mid])
            
        else:
            data = services_member.getMembersData(nwid, [mid])
       
        response_object = Response(200, "ok", data)

    elif request.method == "DELETE":
        services_member.deleteMemberAdditionalData(nwid, mid)

        zerotier_api.make_request(request.method, f"/controller/network/{nwid}/member/{mid}")
        
        defaultConfig = {
            "authorized": False,
            "ipAssignments": [],
            "capabilities": [],
            "tags": []
            }

        zerotier_api.make_request("POST", f"/controller/network/{nwid}/member/{mid}", json = defaultConfig)

        response_object = Response(200, "ok")

        
        
    return jsonify(response_object)